command-line-interface.md

```diff
@@ -0,0 +1,150 @@
+# Operations
+
+A conformant [runtime][] MUST provide an executable (called `funC` in the following examples).
+That executable MUST support commands with the following template:
+
+```sh
+$ funC [global-options] <COMMAND> [command-specific-options] <command-specific-arguments>
+```
+
+## Global options
+
+None are required, but the runtime MAY support options that start with at least one hyphen.
+Global options MAY take positional arguments (e.g. `--log-level debug`).
+Command names MUST NOT start with hyphens.
+The option parsing MUST be such that `funC <COMMAND>` is unambiguously an invocation of `<COMMAND>` (even for commands not specified in this document).
+If the runtime is invoked with an unrecognized command, it MUST exit with a nonzero exit code and MAY log a warning to stderr.
+
+## Character encodings
+
+This API specification does not cover character encodings, but runtimes SHOULD conform to their native operating system.
+For example, POSIX systems define [`LANG` and related environment variables][posix-lang] for [declaring][posix-locale-encoding] [locale-specific character encodings][posix-encoding], so a runtime in an `en_US.UTF-8` locale SHOULD write its [version](#version) to stdout in [UTF-8][].
+
+## Commands
+
+### version
+
+Print the runtime version and exit.
+
+* *Options* None are required, but the runtime MAY support options.
+* *Standard streams*
+    * *stdin:* The runtime MUST NOT attempt to read from its stdin.
+    * *stdout:* The runtime MUST print its name, a space, and its version as the first line to its stdout.
+      The name MAY contain any Unicode characters, but MUST NOT contain control codes or newlines.
+      The runtime MAY print additional lines to its stdout, and the format for those lines is not specified in this document.
+    * *stderr:* The runtime MAY print diagnostic messages to stderr, and the format for those lines is not specified in this document.
+* *Exit code:* The runtime MUST exit with zero.
+
+#### Example
+
+```
+$ funC version
+funC 1.0.0
+Built for x86_64-pc-linux-gnu
+$ echo $?
+0
+```
+
+### start
+
+[Start][start] a container from a [bundle directory][bundle].
+
+* *Options*
+    * *`--id <ID>`* Set the container ID when creating a container.
```

Comment:
On Sun, Jul 10, 2016 at 07:13:27AM -0700, Mike Brown wrote:
> > +### start
> > +
> > +[Start][start] a container from a [bundle directory][bundle].
> > +
> > +* *Options*
> > +    * *`--id <ID>`* Set the container ID when creating a container.
>
> We support multiple instances of a container running at the same
> time in a host.  The model we are using to support that is by
> forcing the user to provide a unique container id in that host
> environment.  Even run/create would not work without a unique
> reference id.

I agree with all of that except that it has to be the user that
generates the ID.  If you have a long-running create that can
automatically delete persistent bits (e.g. cgroups) for you after the
container process exits, then the runtime-caller should be free to
tell the runtime “I don't care what the ID is, pick whatever you
like”.  That doesn't sound like a from-scratch redesign to me ;).

And runC itself supported an optional container ID until February's
opencontainers/runc#541, although I see no motivation for the change
in either that PR or it's backing opencontainers/runc#532.